### ISSUE TITLE
Failed to evaluate variables

### ISSUE THREAD

**Issue by trrahul** [codingseb/ExpressionEvaluator#49]
I used the following code.
```
ExpressionEvaluator evaluator = new ExpressionEvaluator();
            evaluator.Variables = new Dictionary<string, object>();
            evaluator.Variables["a"] = 1;
            evaluator.Variables["b"] = 2;
            evaluator.Variables["c"] = "a+b";
            evaluator.Variables["d"] = "c+3";
            Console.WriteLine(evaluator.Evaluate("a"));
            Console.WriteLine(evaluator.Evaluate("b"));
            Console.WriteLine(evaluator.Evaluate("c"));
            Console.WriteLine(evaluator.Evaluate("d"));
```

The output was
```
1
2
a+b
c+3
```
Shouldn't the output be the following?

```
1
2
3
6
```

**Comment by codingseb** [comment]
Hello @trrahul.
No the output are right.

This because a variable contains a C# object and in your case. variables `c` and `d` contains a `string`. So they are interprated as `string`.
So if you do this :
```c#
evaluator.Evaluate("c+d");
// or if
evaluator.Variables["h"] = "Hello";
evaluator.Variables["w"] = "World";
evaluator.Evaluate("h+\" \" +w");
``` 
you get : 
```
a+bc+3
Hello World
```
but if you want to evaluate a string in a evaluation you could do something like this : 
```c#
evaluator.Evaluate("Evaluate(c)");
evaluator.Evaluate("Evaluate(d)");
```
And then you get : 
```
3
6
```

**Comment by codingseb** [comment]
After a little thought. I think it might be good to introduce a type of variable that could be evaluated when encountered to better answer to this issue.  
Something like : 
```c#
evaluator.Variables["c"] = new SubExpression("a+b");
evaluator.Variables["d"] = new SubExpression("c+3");
```

I Reopen this issue and will work on this for the next version of ExpressionEvaluator.
So @trrahul thanks for this issue.

**Comment by trrahul** [comment]
Isn't it better to search for the variables in the expression and evaluate them in order instead of creating a new subexpression?

**Comment by codingseb** [comment]
You need to understand  that variables are just containers like C# variables or any programming language variables. It could contains anything (string, int, float, Regex, File or a custom class...) and a `string` is not a expression it's a `string` and it's right like this because you can do operation with `string` variables so you need a specific type to specify that the content of a variable must be evaluate otherwise the variable is simply evaluated as it's content.

> Because in this case, please correct me if I am wrong, every variable has to be created as a subexpression if they are used later in another expression.  

No if a variable is just a value like your variables `a` or `b` (int) they are interpreted as this.
So only variables that need to be evaluated when should need to be contains in a specific type that is that ExpressionEvaluator know it' need to be sub evaluate.

**Comment by trrahul** [comment]
Thanks for the clarification. I was trying out different expression evaluators and this issue happened with all the libs I tried except lambdaparser. It was able to evaluate the expression correctly, but it had another problem ExpressionEvaluator did not have. https://github.com/nreco/lambdaparser/issues/22

**Comment by codingseb** [comment]
I just published version 1.4.9.0 that allow to the use of SubExpression here is the doc on the wiki [SubExpression variable](https://github.com/codingseb/ExpressionEvaluator/wiki/Variables-and-Functions#subexpression-variable)
